CLOUDP-305862 - Add missing rule Technical Specifications (#546)
mongodb/openapi: tools/spectral/ipa/rulesets/README.md

##### Implementation details
Rule checks for the following conditions:
  - Applies only to GET methods on resource collection paths
  - Ignores singleton resources
  - Verifies that the operation object does not contain a requestBody property

#### xgen-IPA-105-resource-has-list

 ![warn](https://img.shields.io/badge/warning-yellow) 
APIs must provide a List method for resources.

##### Implementation details
Rule checks for the following conditions:
  - Applies only to resource collection paths
  - Ignores singleton resources
  - Verifies the resource path has a GET method
  - Fails if the resource path does not have a GET method

#### xgen-IPA-105-list-method-response-is-get-method-response

 ![warn](https://img.shields.io/badge/warning-yellow) 
The response body of the List method should consist of the same resource object returned by the Get method.
##### Implementation details Rule checks for the following conditions:
  - Applies only to resource collection paths with JSON content types
  - Ignores singleton resources
  - Ignores responses without a schema or non-paginated responses
  - A response is considered paginated if it has a schema with a 'results' array property

>   - Verifies that the schema of items in the 'results' array matches the schema used in the Get method response
>   - Fails if the Get method doesn't have a schema reference or if the schemas don't match

Reviewer: ```suggestion
  - Verifies that the schema of items in the 'results' array is the same schema used in the Get method response
  - Fails if the Get method doesn't have a schema reference or if the methods don't reference the same schema
```

 ![warn](https://img.shields.io/badge/warning-yellow) 
The Create method request should be a Request suffixed object.

##### Implementation details
Rule checks for the following conditions:

>   - Applies only to POST methods on resource collection paths (non-singleton resources)

Reviewer: ```suggestion
  - Applies only to POST methods on resource collection paths (non-singleton resources and non-custom methods)
```

  - Applies only to JSON content types
  - Verifies the schema references a predefined schema (not inline)
  - Confirms the referenced schema name ends with "Request" suffix

#### xgen-IPA-106-create-method-should-not-have-query-parameters

 ![warn](https://img.shields.io/badge/warning-yellow) 
Create operations should not use query parameters.

##### Implementation details
Rule checks for the following conditions:

>   - Applies only to POST methods on resource collection paths (non-singleton resources)

Reviewer: ```suggestion
  - Applies only to POST methods on resource collection paths (non-singleton resources and non-custom methods)
```

  - Verifies the operation does not contain query parameters

>   - Ignores specified parameters like 'pretty' and 'envelope' via configuration

Reviewer: ```suggestion
  - Ignores specified parameters `pretty` and `envelope` via configuration
```

#### xgen-IPA-106-create-method-request-body-is-get-method-response

 ![warn](https://img.shields.io/badge/warning-yellow) 
Request body content of the Create method and response content of the Get method should refer to the same resource.

##### Implementation details

Validation checks the POST method for resource collection paths.
  - Validation ignores resources without a Get method.
  - `readOnly:true` properties of Get method response will be ignored. 
  - `writeOnly:true` properties of Create method request will be ignored.
  - Property comparison is based on `type` and `name` matching.
  - `oneOf` and `discriminator` definitions must match exactly.

#### xgen-IPA-106-create-method-request-has-no-readonly-fields

 ![warn](https://img.shields.io/badge/warning-yellow) 
Create method Request object must not include fields with readOnly:true.

##### Implementation details
Rule checks for the following conditions:

>   - Applies only to POST methods on resource collection paths (non-singleton resources)

Reviewer: ```suggestion
  - Applies only to POST methods on resource collection paths (non-singleton resources and non-custom methods)
```

  - Applies only to JSON content types

>   - Searches through the schema to find any properties marked with readOnly attribute
>   - Fails if any readOnly properties are found in the request schema

Reviewer: ```suggestion
  - Searches through the schema to find any properties marked with `readOnly` attribute
  - Fails if any `readOnly` properties are found in the request schema
```

#### xgen-IPA-106-create-method-response-code-is-201

 ![warn](https://img.shields.io/badge/warning-yellow) 
Create methods must return a 201 Created response code.

##### Implementation details
Rule checks for the following conditions:

>   - Applies only to POST methods on resource collection paths (non-singleton resources)

Reviewer: ```suggestion
  - Applies only to POST methods on resource collection paths (non-singleton resources and non-custom methods)
```

  - Verifies the 201 Created response code is present
  - Fails if the method lacks a 201 Created response or defines a different 2xx status code

#### xgen-IPA-106-create-method-response-is-get-method-response

 ![warn](https://img.shields.io/badge/warning-yellow) 
The response body of the Create method should consist of the same resource object returned by the Get method.

##### Implementation details
Rule checks for the following conditions:

>   - Applies only to POST methods on resource collection paths

Reviewer: ```suggestion
  - Applies only to POST methods on resource collection paths (non-singleton resources and non-custom methods)
```

 ![warn](https://img.shields.io/badge/warning-yellow) 
Delete method response should not have schema reference to object.

##### Implementation details
Rule checks for the following conditions:
  - Applies to 204 responses in DELETE methods
  - Verifies that the response does not contain a schema property
  - Fails if any content type in the response has a defined schema as reference

#### xgen-IPA-108-delete-method-return-204-response

 ![warn](https://img.shields.io/badge/warning-yellow) 

> DELETE method must return 204 No Content.

Reviewer: ```suggestion
Delete methods must return 204 No Content.
```

##### Implementation details
Rule checks for the following conditions:
  - Applies to all DELETE methods
  - Verifies that the method includes a 204 response code
  - Ensures no other 2xx response codes are defined
  - Fails if the 204 status code is missing or if other 2xx responses exist

#### xgen-IPA-108-delete-include-404-response

 ![warn](https://img.shields.io/badge/warning-yellow) 

> DELETE method must include 404 response and return it when resource not found.

Reviewer: ```suggestion
Delete methods must include 404 response and return it when resource not found.
```

Author: Ok. So there is no sense for us to use official http vocabulary. Noted

Reviewer: At least in the summary, in the implementation details we can go into HTTP methods and stuff like that 👍

##### Implementation details
Rule checks for the following conditions:
  - Applies to all DELETE methods
  - Verifies that the method includes a 404 response code
  - Fails if the 404 status code is missing from the responses

#### xgen-IPA-108-delete-request-no-body

 ![warn](https://img.shields.io/badge/warning-yellow) 

> DELETE method must not have request body.

Reviewer: ```suggestion
Delete methods must not have request body.
```

 ![warn](https://img.shields.io/badge/warning-yellow) 
Singleton resources must not have a user-provided or system-generated ID.

##### Implementation details
Rule checks for the following conditions:
  - Applies only to singleton resources that are identified as resource collection identifiers
  - Checks that the resource has a GET method defined
  - Examines all 2xx response schemas from the GET method

>   - Verifies that no schema contains 'id' or '_id' properties in their object definitions

Reviewer: ```suggestion
  - Verifies that no schema contains `id` or `_id` properties in their object definitions
```
